demos/dev_mode_key_provisioning/src/aws_dev_mode_key_provisioning.c

```diff
@@ -868,51 +910,59 @@ void vAlternateKeyProvisioning( ProvisioningParams_t * xParams )
-            xDestroyCredentials( xSession );
-        #endif
-
         xResult = xProvisionDevice( xSession, xParams );
 
         pxFunctionList->C_CloseSession( xSession );
     }
+
+    return xResult;
 }
 /*-----------------------------------------------------------*/
 
-void vDevModeKeyProvisioning( void )
+CK_RV vDevModeKeyProvisioning( void )
 {
     ProvisioningParams_t xParams;
 
+    xParams.pucJITPCertificate = ( uint8_t * ) keyJITR_DEVICE_CERTIFICATE_AUTHORITY_PEM;
     xParams.pucClientPrivateKey = ( uint8_t * ) keyCLIENT_PRIVATE_KEY_PEM;
     xParams.pucClientCertificate = ( uint8_t * ) keyCLIENT_CERTIFICATE_PEM;
-    xParams.pucJITPCertificate = ( uint8_t * ) keyJITR_DEVICE_CERTIFICATE_AUTHORITY_PEM;
 
     /* If using a JITR flow, a JITR certificate must be supplied. If using credentials generated by
      * AWS, this certificate is not needed. */
-    if( ( NULL == xParams.pucJITPCertificate ) ||
-        ( 0 == strcmp( "", ( const char * ) xParams.pucJITPCertificate ) ) )
+    if( ( NULL != xParams.pucJITPCertificate ) &&
+        ( 0 != strcmp( "", ( const char * ) xParams.pucJITPCertificate ) ) )
     {
-        xParams.ulJITPCertifiateLength = 0;
+        xParams.ulJITPCertificateLength = 1 + strlen( ( const char * ) xParams.pucJITPCertificate );
     }
     else
     {
-        xParams.ulJITPCertifiateLength = 1 + strlen( ( const char * ) xParams.pucJITPCertificate );
+        xParams.pucJITPCertificate = NULL;
     }
 
-    if( ( NULL == xParams.pucClientPrivateKey ) ||
-        ( 0 == strcmp( "", ( const char * ) xParams.pucClientPrivateKey ) ) ||
-        ( NULL == xParams.pucClientCertificate ) ||
-        ( 0 == strcmp( "", ( const char * ) xParams.pucClientCertificate ) ) )
+    /* The hard-coded client certificate and private key can be useful for
+     * first-time lab testing. They are optional after the first run, though, and
+     * not recommended at all for going into production. */
+    if( ( NULL != xParams.pucClientPrivateKey ) &&
+        ( 0 != strcmp( "", ( const char * ) xParams.pucClientPrivateKey ) ) )
     {
-        configPRINTF( ( "ERROR: the vDevModeKeyProvisioning function requires a valid device private key and certificate.\r\n" ) );
-        configASSERT( pdFALSE );
+        xParams.ulClientPrivateKeyLength = 1 + strlen( ( const char * ) xParams.pucClientPrivateKey );
     }
     else
     {
-        xParams.ulClientPrivateKeyLength = 1 + strlen( ( const char * ) xParams.pucClientPrivateKey );
+        xParams.pucClientPrivateKey = NULL;
+    }
+
+    if( ( NULL != xParams.pucClientCertificate ) ||
+        ( 0 != strcmp( "", ( const char * ) xParams.pucClientCertificate ) ) )
+    {
         xParams.ulClientCertificateLength = 1 + strlen( ( const char * ) xParams.pucClientCertificate );
-        vAlternateKeyProvisioning( &xParams );
     }
+    else
+    {
+        xParams.pucClientCertificate = NULL;
+    }
+
+    return vAlternateKeyProvisioning( &xParams );
```

Comment:
Ok so the behavior now is just to go ahead and provision always and then the application fail TLS if one of the keys is NULL.

Reply:
The other changes above will do a better job of detecting that now, including in scenarios where the customer is trying to use a pre-provisioned part.